pkg/build/buildkit/k8s_autodiscovery.go

```diff
@@ -0,0 +1,351 @@
+
+	case err = <-errCh:
+		releaseLeaderLock("") // release all lease locks
+		return nil, err
+
+	case pod = <-selected:
+	}
+
+	defer releaseLeaderLock(pod.Name) // release all lease locks except from the choosen pod
+
+	return pod, nil
+}
+
+func (d *k8sDiscoverer) buildkitPodNamespace(ctx context.Context, opts KubernertesDiscoveryOptions, app string) (string, error) {
+	if !opts.UseSameNamespaceAsApp {
+		return opts.Namespace, nil
+	}
+
+	klog.V(4).Infof("Discovering the namespace where app %s is running on...", app)
+
+	tsuruApp, err := d.dcs.Resource(tsuruAppGVR).Namespace(TsuruAppNamespace).Get(ctx, app, metav1.GetOptions{})
+	if err != nil {
+		return "", err
+	}
+
+	// See more about App resource at: https://github.com/tsuru/tsuru/blob/main/provision/kubernetes/pkg/apis/tsuru/v1/types.go#L24
+	ns, found, err := unstructured.NestedString(tsuruApp.Object, "spec", "namespaceName")
+	if err != nil {
+		return "", err
+	}
+
+	if !found {
+		return "", fmt.Errorf("failed to fetch namespace in the App resource")
+	}
+
+	klog.V(4).Infof("App %s is running on namespace %s...", app, ns)
+
+	return ns, nil
+}
+
+func watchBuildKitPods(ctx context.Context, cs *kubernetes.Clientset, labelSelector, ns string, pods chan<- *corev1.Pod) error {
+	w, err := cs.CoreV1().Pods(ns).Watch(ctx, metav1.ListOptions{
+		LabelSelector: labelSelector,
+		Watch:         true,
+	})
+	if err != nil {
+		return err
+	}
+
+	defer w.Stop()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+
+		case e := <-w.ResultChan():
+			if e.Type != watch.Added && e.Type != watch.Modified {
+				continue
+			}
+
+			pod := e.Object.(*corev1.Pod)
+			if isPodReady(pod) {
+				pods <- pod
+			}
+
+			klog.V(4).Infof("Pod %s/%s is not ready yet", pod.Namespace, pod.Name)
+		}
+	}
+}
+
+func acquireLeaseForPod(ctx context.Context, cs *kubernetes.Clientset, ch chan<- *corev1.Pod, pod *corev1.Pod, opts KubernertesDiscoveryOptions) {
+	podname := os.Getenv("POD_NAME")
+	if podname == "" {
+		podname, _ = os.Hostname()
+	}
+
+	uniqueHolderName := fmt.Sprintf("%s-%d", podname, time.Now().Unix())
+
+	klog.V(4).Infof("Attempting to acquire the lease for pod %s/%s under holder name %q...", pod.Namespace, pod.Name, uniqueHolderName)
+
+	leaderelection.RunOrDie(ctx, leaderelection.LeaderElectionConfig{
+		Lock: &resourcelock.LeaseLock{
+			LeaseMeta: metav1.ObjectMeta{
+				Name:      fmt.Sprintf("%s-%s", strings.TrimRight(opts.LeasePrefix, "-"), pod.Name),
+				Namespace: pod.Namespace,
+			},
+			Client: cs.CoordinationV1(),
+			LockConfig: resourcelock.ResourceLockConfig{
+				Identity: uniqueHolderName,
+			},
+		},
+		ReleaseOnCancel: true,
+		LeaseDuration:   5 * time.Second,
+		RenewDeadline:   2 * time.Second,
+		RetryPeriod:     500 * time.Millisecond,
+		Callbacks: leaderelection.LeaderCallbacks{
+			OnStartedLeading: func(c context.Context) {
+				select {
+				case ch <- pod:
+					klog.V(4).Infof("Selected BuildKit pod: %s/%s", pod.Namespace, pod.Name)
```

Comment:
adding the `uniqueHolderName` on this log could be useful (also, the 2 logs after this one)

Reply:
Honestly, this info (uniqueHolderName) doesn't help anything. This name it's mostly like a random number around each build call, it's important just to avoid conflict while getting the lease (lock). Two consecutive build calls will generate different holder names, so they will not share the lease lock.